client/allocrunner/alloc_runner.go

```diff
@@ -553,8 +553,9 @@ func (ar *allocRunner) clientAlloc(taskStates map[string]*structs.TaskState) *st
 
 		// If we are part of a deployment and the task has failed, mark the
 		// alloc as unhealthy. This guards against the watcher not be started.
+		// If the health status is already set then terminal allocations should not
 		if a.ClientStatus == structs.AllocClientStatusFailed &&
-			alloc.DeploymentID != "" && !a.DeploymentStatus.IsUnhealthy() {
+			alloc.DeploymentID != "" && !a.DeploymentStatus.HasHealth() && !a.DeploymentStatus.IsUnhealthy() {
```

Comment:
```suggestion
			alloc.DeploymentID != "" && !a.DeploymentStatus.HasHealth() {
```

If the alloc _does not have a health status_ it cannot be healthy or unset (`!a.DeploymentStatus.IsUnhealthy()`).